File: built-in plugins dsp/mb-sampler-piano.cpp
/**
 * MB Piano Sampler
 * Category : instrument
 * Type     : sampler
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multi-sampled grand piano
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SAMPLER_PIANO_H
#define MB_SAMPLER_PIANO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSamplerPiano : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-sampler-piano";
    static constexpr const char* PLUGIN_NAME    = "MB Piano Sampler";
    static constexpr const char* PLUGIN_TYPE    = "sampler";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float velocity = 0.8f;  // range [0, 1]
    float release = 0.5f;  // range [0.1, 5]
    float volume = 0.8f;  // range [0, 1]
    };

    MbSamplerPiano() = default;
    ~MbSamplerPiano() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.velocity = std::clamp(params.velocity, 0f, 1f);
        params.release = std::clamp(params.release, 0.1f, 5f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Piano Sampler
        return input;
    }
};

#endif // MB_SAMPLER_PIANO_H

File: built-in plugins dsp/mb-sampler-piano.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSamplerPianoPlugin: PluginDefinition = { id: 'mb-sampler-piano', slug: 'mb-sampler-piano', name: 'MB Piano Sampler', category: 'instrument', type: 'sampler', version: '1.0.0', description: 'Multi-sampled grand piano', author: 'Max Booster', oscillators: [], envelope: { attack: 0.001, decay: 0.5, sustain: 0.8, release: 0.5 }, parameters: [{ id: 'velocity', name: 'Velocity', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 0.5, minValue: 0.1, maxValue: 5, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { velocity: 0.8, release: 0.5, volume: 0.8 } };

export default MbSamplerPianoPlugin;
